\documentclass[a4paper,10pt]{article}
\usepackage[utf8]{inputenc}
\usepackage{natbib}
\bibliographystyle{unsrtnat}
%opening
\title{n-gram analysis of biological sequences in R}
\author{}
\date{}

\begin{document}

\maketitle

n-grams (k-mers) are subvectors of $n$ characters derived from input sequences. 
Since n-grams capture positional and compositional properties of a sequence, 
they are widely used in genomics, transcriptomics and proteomics. Despite the 
continuous interest, there are only few tools tailored for comparative n-gram 
studies. Furthermore, the volume of n-gram data is usually very large, making 
its analysis in \textbf{R} especially challenging. 

The CRAN package \textit{biogram}~\citep{burdukiewicz_biogram:_2015} facilitates 
incorporating n-gram data in the \textbf{R} workflows. Aside from the efficient 
extraction and storage of n-grams, the package offers also a novel feature 
selection method designed specifically for n-gram data. QuiPT (Quick 
Permutation Test) uses filtering criteria such as information gain 
(mutual information) to choose the most informative n-grams. To speed up the 
computation and allow precise estimation of small p-values, QuiPT uses 
analytically derived distributions instead of a large number of permutations. In 
addition to this, \textit{biogram} contains tools designed for reducing the 
dimensionality of the amino acid alphabet~\citep{murphy_simplified_2000}, 
further scaling down the feature space.

To illustrate the usage of n-gram data in the analysis of biological sequences,  
we present two case studies performed solely in \textbf{R}. The first,   
prediction of amyloids, short proteins associated with the number of clinical   
disorders as Alzheimer's or Creutzfeldt-Jakobâ€™s 
diseases~\citep{fandrich_oligomeric_2012}, employs random 
forests~\citep{wright_ranger:_2015} trained on n-grams. The second, detection   
of signal peptides orchestrating an extracellular transport of proteins, 
utilizes more complicated probabilistic framework (Hidden semi-Markov 
model) based on the n-gram frequency. 

\bibliography{amyloids}

\end{document}
